lib/spack/spack/installer.py

```diff
@@ -419,18 +421,29 @@ def _try_install_from_binary_cache(pkg, explicit, unsigned=False, timer=timer.NU
             otherwise, ``False``
         timer (Timer):
     """
+    # Early exit if no mirrors are configured.
+    if not spack.mirror.MirrorCollection():
+        return False
+
     pkg_id = package_id(pkg)
     tty.debug("Searching for binary cache of {0}".format(pkg_id))
 
     timer.start("search")
-    matches = binary_distribution.get_mirrors_for_spec(pkg.spec)
+    # Note: this is an optimization to order mirrors _before_ fetching, therefore it's
+    # offline only.
+    prioritized_fetch_urls = [
+        hit["mirror_url"]
+        for hit in binary_distribution.get_mirrors_for_spec(pkg.spec, index_only=True)
```

Comment:
Trying to think this bit through.  Seems we'll continue to check this out of a sense of optimism, but under the hood make no requests in case of a miss.  Is there any chance we'll get a hit without having arranged to call `bindist.update()` in the current process first?  At one point I thought we should just remove this (see #31347), but I guess if it's cheap, it doesn't hurt anything.

Reply:
It can still be useful in the "normal" use case of `spack concretize` + `spack install`, since concretization with reuse will fetch the index.json, so a few seconds later during install the cache can be expected to be up to date and will likely lead to direct binary cache hits.